Sources/Linq2DynamoDb.DataContext/TableDefinitionWrapper.cs

```diff
@@ -360,45 +369,120 @@ protected override void InitReader(ISupervisableEnumerable reader)
+                        this.Cache.RemoveEntities(addedEntities.Keys.Union(modifiedEntities.Keys).Union(removedEntities.Keys));
 
                         throw updateTask.Exception;
                     }
+
+                    if (!string.IsNullOrEmpty(VersionPropertyName))
+                    {
+                        // Update the cache if there is a Version Property. Version properties change 
+                        // during insertion/updating so we want to have the latest values for versions 
+                        // of added and modified entities. 
+                        this.Cache.UpdateCacheAndIndexes(addedEntities, modifiedEntities, removedEntities.Keys);
+                    }
                 }
             );
         }
 
+        /// <summary>
+        /// Performs create/update operations with DynamoDB, respecting the [DynamoDBVersion] attribute if present
+        /// </summary>
+        /// <remarks>
+        /// If the entity is new (Version property has not been set) set the version property 0. If 
+        /// updating an existing entity increment the Version property by 1.
+        /// </remarks>
+        /// <param name="entitiesToAdd">Dictionary of entity keys and documents to add/update</param>
+        /// <returns>An IEnumerable containing one task for each entity to add or update.</returns>
+        private IEnumerable<Task> AddEntriesAsync(IDictionary<EntityKey, Document> entitiesToAdd) 
+        {
+            return entitiesToAdd
+                .Values
+                .Select(entry => {
+                    var putConfiguration = new PutItemOperationConfig();
+                    if (!string.IsNullOrEmpty(VersionPropertyName)) 
+                    {
+                        if (entry[VersionPropertyName] != null && 
+                            entry[VersionPropertyName].ToPrimitive(typeof(object)).Value == null)
+                        {
+                            entry[VersionPropertyName] = 0;
+
+                            putConfiguration.ConditionalExpression = new Expression() 
+                            {
+                                ExpressionStatement = $"attribute_not_exists(#key)",
+                                ExpressionAttributeNames = new Dictionary<string, string>
+                                {
+                                    { "#key", HashKeyPropetyName }
+                                }
+                            };
+                        }
+                        else
+                        {
+                            putConfiguration.Expected = new Document(new Dictionary<string, DynamoDBEntry>()
+                            {
+                                { VersionPropertyName, entry[VersionPropertyName] }
+                            });
+
+                            // TODO: We assume that the Version is an int here but it could be 
```

Comment:
This is a lazy approach. The correct way to do this is to check the type of the object's versioning property, call the appropriate conversion method on the `DynamoDBEntry` (e.g. `AsSingle()` in the case of a float, `AsInt()` in the case of an int, etc.), then put the incremented value back in the Document object. 

In the current implementation, if the version property were a `float` and the version were stored as 1.1 (by some external process) this code would overwrite that value to 2 instead of 2.1. This doesn't break optimistic locking behavior, but it could produce unexpected results for anyone who is checking the value of the version property.